### PR TITLE
Fixes #2388. Add tests for the new method/setter rules. Update assertions

### LINKED ISSUE #2388: [Extension types] Update tests according to the new setter/method rules
Update Extension types tests according to the new rules regarding setter/method conflict. The [change](https://github.com/dart-lang/language/commit/299d20be62cd23fc396d6fa812d6e7d92386ab09)

### PR DESCRIPTION
**Pull request by sgrekhov**

### CODE REVIEW

File: LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t01.dart [hunk 1]
```diff
@@ -2,11 +2,12 @@
...
-/// whose interface contains a member signature named n.
+/// @assertion We say that an extension type declaration DV has an extension
+/// type member named n in the cases where:
+/// - DV declares a member named n.
+/// - DV has no such declaration, but DV has a direct extension type
+///   superinterface V that has an extension type member named n due to a member
+///   declaration DM2, and DV does not declare a member that precludes DM2.
 ///
 /// @description Checks that if an extension type `ET` has a superinterface with
 /// a member `m` then this member is also presents in `ET`
```
Comment (eernstg) [inline LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t01.dart:13]: I think we have a number of occurrences of this:
```suggestion
/// a member `m` then this member is also present in `ET`
```

File: LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t01.dart [hunk 2]
```diff
@@ -2,11 +2,12 @@
...
-/// type member named n in the case where DV does not declare a member named n,
-/// but DV has a direct extension type superinterface V that has a non-extension
-/// type member named n, or DV has a direct non-extension type superinterface T
-/// whose interface contains a member signature named n.
+/// @assertion We say that an extension type declaration DV has an extension
+/// type member named n in the cases where:
+/// - DV declares a member named n.
+/// - DV has no such declaration, but DV has a direct extension type
+///   superinterface V that has an extension type member named n due to a member
+///   declaration DM2, and DV does not declare a member that precludes DM2.
```
Comment (eernstg) [inline LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t01.dart:10]: Right, this is actually testing a situation where the "inherited" member is an extension type member.

File: LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t02.dart [hunk 3]
```diff
@@ -2,11 +2,12 @@
...
-/// type member named n in the case where DV does not declare a member named n,
-/// but DV has a direct extension type superinterface V that has a non-extension
-/// type member named n, or DV has a direct non-extension type superinterface T
-/// whose interface contains a member signature named n.
+/// @assertion We say that an extension type declaration DV has an extension
+/// type member named n in the cases where:
+/// - DV declares a member named n.
+/// - DV has no such declaration, but DV has a direct extension type
+///   superinterface V that has an extension type member named n due to a member
+///   declaration DM2, and DV does not declare a member that precludes DM2.
```
Comment (eernstg) [inline LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t02.dart:10]: In this case the member is indeed a non-extension type member of the extension type, so the `@assertion` should probably be taken from the paragraph right after the one about extension type members which is the current `@assertion`.

Comment (eernstg) [reply LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t02.dart:10]: Please check the other tests for the same issue: The `@assertion` should be about extension type members when the test is about extension type members, and similarly for non-extension type members.

File: LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t06.dart [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+extension type V2(int id) {
+  String get n => "V2";
+}
+
+extension type V3(int id) {
+  void set n(int n) {}
+}
+
+extension type ET1(int id) implements V1 {
+  T n<T>(T t) => t;
```
Comment (eernstg) [inline LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t06.dart:33]: This one doesn't actually preclude anything: `V1.n` is just redeclared.

File: LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t06.dart [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+extension type ET2(int id) implements V2 {
+  T n<T>(T t) => t;
```
Comment (eernstg) [inline LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t06.dart:37]: This one doesn't actually preclude anything: `V2.n` is just redeclared.

File: LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t06.dart [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+extension type ET3(int id) implements V3 {
+  T n<T>(T t) => t;
```
Comment (eernstg) [inline LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t06.dart:41]: Right, this one precludes `V3.n=`!

Also, `ET4` fits the same description.

File: LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t07.dart [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+extension type ET2(int id) implements V2 {
+  void set n(String v) {
+    log += v;
+  }
+}
+
+extension type ET3(int id) implements V3 {
+  void set n(String v) {
+    log += v;
+  }
```
Comment (eernstg) [inline LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t07.dart:49]: This one doesn't preclude anything: `V3.n=` is just redeclared.

File: LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t07.dart [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+extension type ET4(int id) implements V1, V2, V3 {
+  void set n(String v) {
+    log += v;
+  }
```
Comment (eernstg) [inline LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t07.dart:55]: Right, this one precludes `V1.n` and `V2.n`, and it redeclares `V3.n=`.

**Review by eernstg — COMMENTED** [review]
Looks good! However, there are some typos, and the `@assertion` talks about a different thing than the thing which is actually tested in a couple of cases.

**Comment by sgrekhov** [comment]
Tests updated. Changes
-  tests for extension-type members and non-extension type members are separated
- removed tests that check redeclaration instead of precluding
- `static_analysis_member_invocation_A06_t*.dart` will re reordered after the merge in a separate PR

File: LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t02.dart (hunk 3 above)
Comment (eernstg) [reply LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t02.dart]: Thanks!

**Review by eernstg — COMMENTED** [review]
Looks good! I think a couple of declarations are unused. Also, some future reader is probably going to be happy if class names and extension type names follow the conventions (so `V...` typically means "a helper declaration whose kind is `extension type`", not "a class").

File: LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t03.dart [hunk 9]
```diff
@@ -25,18 +26,6 @@ extension type ET1(int id) implements ET0 {}
 
 extension type ET2(int id) implements num {}
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LanguageFeatures/Extension-types/static_analysis_member_invocation_A06_t03.dart]: Yes, of course. Thank you!

File: LanguageFeatures/Extension-types/static_analysis_member_invocation_A08_t02.dart [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+///   precludes m.
+///
+/// @description Checks that if an extension type `ET` has a superinterface with
+/// a member `m` then this member is also present in `ET` but members of its
+/// representation type aren't
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=inline-class
+
+extension type ET0(int id) {
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LanguageFeatures/Extension-types/static_analysis_member_invocation_A08_t02.dart]: Yes. Removed

File: LanguageFeatures/Extension-types/static_analysis_member_invocation_A08_t04.dart [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+/// @description Checks that a getter doesn't preclude setter and vice versa
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=inline-class
+
+import "../../Utils/expect.dart";
+
+String log = "";
+
+class V1 {
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LanguageFeatures/Extension-types/static_analysis_member_invocation_A08_t04.dart]: Agreed. Probably it is not so tiny thing. Naming is important! Renamed

**Review by sgrekhov — COMMENTED** [review]
Updated. Unused code removed, classes renamed

File: LanguageFeatures/Extension-types/static_analysis_member_invocation_A08_t04.dart (hunk 11 above)
[the remark this answers is not in the document]
Comment (eernstg) [reply LanguageFeatures/Extension-types/static_analysis_member_invocation_A08_t04.dart]: Thanks!

**Review by eernstg — APPROVED** [review]
LGTM!

**Pull request merged**